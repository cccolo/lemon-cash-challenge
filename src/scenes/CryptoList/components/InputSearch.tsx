import React from 'react';
import {StyleSheet, TextInput, TouchableOpacity, View} from 'react-native';
import _ from 'lodash';

import {CloseIcon, SearchIcon} from '../../../assets/images';

type Props = {
  onCancel: () => void;
  onChangeText: (value: string) => void;
};

export const InputSearch: React.FC<Props> = ({
  onCancel,
  onChangeText,
}: Props) => {
  const [searchQuery, setSearchQuery] = React.useState('');

  const setOnChangeTextDebounced = React.useRef(
    _.debounce((text: string) => onChangeText(text), 500),
  );

  const handleSearch = (text: string) => {
    setSearchQuery(text);
    setOnChangeTextDebounced.current(text);
  };

  return (
    <View style={styles.searchContainer}>
      <View style={styles.searchInputContainer}>
        <SearchIcon />
        <TextInput
          style={styles.searchInput}
          placeholder="Ingresa nombre o simbolo"
          placeholderTextColor="white"
          value={searchQuery}
          onChangeText={handleSearch}
        />
        <View style={styles.searchInputCancelContainer}>
          <TouchableOpacity
            onPress={() => {
              setSearchQuery('');
              onCancel();
            }}>
            <CloseIcon style={styles.searchInputCancel} />
          </TouchableOpacity>
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  searchContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  searchInputContainer: {
    alignItems: 'center',
    borderColor: 'white',
    borderRadius: 5,
    borderWidth: 1,
    color: 'white',
    flex: 1,
    flexDirection: 'row',
    height: 40,
    marginBottom: 10,
    paddingLeft: 10,
  },
  searchInput: {
    color: 'white',
    maxWidth: 260,
    paddingLeft: 10,
  },
  searchInputCancelContainer: {
    flex: 1,
    alignItems: 'flex-end',
    marginRight: 12,
  },
  searchInputCancel: {
    color: 'white',
    textAlign: 'right',
  },
});
