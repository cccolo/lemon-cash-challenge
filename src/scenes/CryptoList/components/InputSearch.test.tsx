import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {InputSearch} from './InputSearch';

describe('InputSearch Component', () => {
  const onCancelMock = jest.fn();
  const onChangeTextMock = jest.fn();

  it('renders correctly and matches snapshot', () => {
    const {toJSON} = render(
      <InputSearch onCancel={onCancelMock} onChangeText={onChangeTextMock} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onChangeText when text is changed', () => {
    const {getByTestId} = render(
      <InputSearch onCancel={onCancelMock} onChangeText={onChangeTextMock} />,
    );

    const searchInput = getByTestId('input');
    fireEvent.changeText(searchInput, 'Bitcoin');
    expect(searchInput.props.value).toBe('Bitcoin');
  });

  it('clears search query and calls onCancel when cancel button is pressed', () => {
    const {getByTestId} = render(
      <InputSearch onCancel={onCancelMock} onChangeText={onChangeTextMock} />,
    );

    const cancelButton = getByTestId('cancel-button');
    const searchInput = getByTestId('input');
    fireEvent.press(cancelButton);
    expect(onCancelMock).toHaveBeenCalled();
    expect(searchInput.props.value).toBe('');
  });
});
