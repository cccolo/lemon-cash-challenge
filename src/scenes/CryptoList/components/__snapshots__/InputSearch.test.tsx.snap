// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch Component renders correctly and matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "white",
        "borderRadius": 5,
        "borderWidth": 1,
        "color": "white",
        "flex": 1,
        "flexDirection": "row",
        "height": 40,
        "marginBottom": 10,
        "paddingLeft": 10,
      }
    }
  >
    <svgMock />
    <TextInput
      onChangeText={[Function]}
      placeholder="Ingresa nombre o simbolo"
      placeholderTextColor="white"
      style={
        {
          "color": "white",
          "maxWidth": 260,
          "paddingLeft": 10,
        }
      }
      testID="input"
      value=""
    />
    <View
      style={
        {
          "alignItems": "flex-end",
          "flex": 1,
          "marginRight": 12,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="cancel-button"
      >
        <svgMock
          style={
            {
              "color": "white",
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
